espr/src/lib.rs

```diff
@@ -1,7 +1,110 @@
 //!
-//! EXPRESS language standardized as [ISO-10303-11](https://www.iso.org/standard/38047.html).
+//! Overview
+//! --------
+//!
+//! ```text
+//! ┌────────────────┐
+//! │ EXPRESS Schema │
+//! └──┬─────────────┘
+//!    │ Tokenize
+//! ┌──▼─────────────────────────┐
+//! │ Abstract Syntax Tree (AST) │
+//! └──┬─────────────────────────┘
+//!    │ Legalize
+//! ┌──▼───────────────────────────────┐
+//! │ Intermediate Representation (IR) │
+//! └──┬───────────────────────────────┘ │
+//!    │ Code Generation                 │ espr's responsibility
+//! ┌──▼────────────────────────┐        ▼
+//! │ Rust code with proc-macro │  ─────────────────────────────────
+//! └──┬────────────────────────┘
+//!    │ proc-macro
+//! ┌──▼──────────────┐ call ┌──────────┐
+//! │ Final Rust code ├──────► ruststep │
+//! └─────────────────┘      └──────────┘
+//! ```
+//!
+//! - Tokenize
+//!   - Read EXPRESS input and parse into abstract syntax tree (AST)
+//!   - [ast] module defines AST structs
+//!   - [parser] module defines parser combinator for parsing EXPRESS language
+//! - Legalize
+//!   - Convert AST to IR (intermediate representation) to ready the following code generation
+//!   - [ir] module defines IR structs, and they implements [ir::Legalize] trait for legalizing from AST
+//! - Code Generation
+//!   - [codegen::rust] module generates Rust code from IR
+//!
+//! Introduction to STEP
+//! ---------------------
+//!
+//! As we describe in [README](https://github.com/ricosjp/ruststep#readme),
+//! STEP consists of three components:
+//!
+//! - Schema language called EXPRESS
+//! - Data format called "exchange structure" or "STEP file"
+//! - Schemas, e.g. AP203 for CAD components
+//!
+//! This short section shows basic concept of EXPRESS schemas and exchange structures
+//! for developers new to STEP ecosystem.
+//!
+//! EXPRESS language, standardized as [ISO-10303-11](https://www.iso.org/standard/38047.html),
+//! is a schema language as a part of STEP ecosystem.
+//!
+//! ```text
+//! SCHEMA my_first_schema;
+//!   ENTITY a;
+//!     x: REAL;
+//!     y: REAL;
+//!   END_ENTITY;
+//!
+//!   ENTITY b;
+//!     z: REAL;
+//!     a: a;
+//!   END_ENTITY;
+//! END_SCHEMA;
+//! ```
+//!
+//! This example schema written in EXPRESS language defines
+//! `my_first_schema` schema and two entities `a` and `b`.
+//! Entity is a collection of primitive types or other entities similar to `struct` in Rust.
+//!
+//! Actual data will be stored in another data format called "exchange structure" or "STEP file",
+//! standardized as [ISO-10303-21](https://www.iso.org/standard/63141.html).
+//! Exchange structure consists of following sections:
+//!
+//! - HEADER
+//! - ANCHOR (optional)
+//! - REFERENCE (optional)
+//! - DATA
+//! - SIGNATURE (optional)
+//!
+//! A data section contains indexed instances:
+//!
+//! ```text
+//! DATA;
+//!   #1 = A(1.0, 2.0);
+//!   #2 = B(3.0, A((4.0, 5.0)));
```

Comment:
inline struct should use `A(4.0, 5.0)`

Reply:
As far as we believe ISO-10303-21,

```
PARAMETER_LIST     = PARAMETER { "," PARAMETER } .
PARAMETER          = TYPED_PARAMETER  |
                     UNTYPED_PARAMETER | OMITTED_PARAMETER  .
TYPED_PARAMETER    = KEYWORD "(" PARAMETER ")" .
UNTYPED_PARAMETER  = "$" | INTEGER | REAL | STRING | RHS_OCCURENCE_NAME
                     | ENUMERATION | BINARY | LIST .
OMITTED_PARAMETER  = "*" .
LIST               = "(" [ PARAMETER { "," PARAMETER } ] ")" .

DATA_SECTION       = "DATA" [ "(" PARAMETER_LIST ")" ] ";" 
                     ENTITY_INSTANCE_LIST "ENDSEC;" .
ENTITY_INSTANCE_LIST = { ENTITY_INSTANCE } .
ENTITY_INSTANCE    = SIMPLE_ENTITY_INSTANCE | COMPLEX_ENTITY_INSTANCE .
SIMPLE_ENTITY_INSTANCE  = ENTITY_INSTANCE_NAME "=" SIMPLE_RECORD ";" .
COMPLEX_ENTITY_INSTANCE = ENTITY_INSTANCE_NAME "=" SUBSUPER_RECORD ";" .
SIMPLE_RECORD      = KEYWORD "(" [ PARAMETER_LIST ] ")" .
SUBSUPER_RECORD    = "(" SIMPLE_RECORD_LIST ")" .
SIMPLE_RECORD_LIST = SIMPLE_RECORD { SIMPLE_RECORD } .
SIGNATURE_SECTION  = "SIGNATURE" SIGNATURE_CONTENT "ENDSEC;".
```

Inline typed parameter must have single parameter:

```
B(3.0, A((4.0, 5.0)));
^ simple record
  ^ parameter
       ^ parameter = typed parameter, this cannot be parameter list
         ^ list
          ^ float
               ^ float
```